src/BatteryStateOfChargeService.cpp

```diff
@@ -35,29 +39,58 @@ namespace
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+ AmpHoursUsed_=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/100);
 }
-
 BatteryStateOfChargeService::~BatteryStateOfChargeService()
 {
 }
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return (AmpHoursUsed_);
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(newCurrent_>=0)
     return false;
+    else
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+  double timeLeft;
+  timeLeft=(AmpHoursUsed_/newCurrent_);
+  timeLeft=qAbs(timeLeft);
+  timeLeft=timeLeft/2.77778e-7;
+
+  QTime base(0,0,0);
+  QTime timeOutput= base.addMSecs(timeLeft);
+  return timeOutput;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+ //Setting Values
+
+    inCurrent_=newCurrent_;
+    initialTime_=currentTime_;
+
+    newCurrent_=batteryData.current;
+    currentTime_=batteryData.time;
+
+    if(firstRun_!=true){
+
+    int mschangeTime=abs(initialTime_.msecsTo(currentTime_));
+    double changeTime=mschangeTime*2.77778e-7;
+
+    double avgCurrent=(newCurrent_+inCurrent_)/2;
+    double pointAverageAH= avgCurrent*changeTime;
+
+    AmpHoursUsed_=AmpHoursUsed_-pointAverageAH;
+    }
+    firstRun_=false;
```

Comment:
Since this line of code `firstRun_ = false;` only needs to be run the first time through, it would be better to include it in the `if` statement. Having it outside implies that some other parts of the code could edit this variable to `true`, which is not true. It also will increase the speed of the program a teeny tiny bit (this last point is not really important since speed is not really an issue with this program).